source/agora/consensus/validation/Transaction.d

```diff
@@ -122,6 +128,13 @@ public string isInvalidReason (
         // current limitation: if any UTXO is frozen, they all must be frozen
         if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
             return "Transaction: Rejected combined inputs (freeze & payment)";
+
+        // current limitation: frozen UTXO cannot be used for data payload.
+        if ((tx.payload.data.length != 0) && (count_freeze > 0))
```

Comment:
No this check is wrong. If we are spending a melted transaction we can do whatever we want with the payment transaction.

The rules should be:

Payment: Can have data payload
Payment which spends a melted freeze: Can have a data payload
Frozen: Cannot have data payload

So your Freeze and Payment checks are inverted in this PR.

Reply:
Okay Good

Reply:
Thank you for review, I fixed all